File: BOJ/Binary_Search/BJ2110.cpp
#include <iostream>
#include <algorithm>

using namespace std;

int n, c, arr[200002];

int main() {
    ios::sync_with_stdio(0);
    cin.tie(0);

    cin >> n >> c;

    for (int i = 0; i < n; i++)
        cin >> arr[i];

    sort(arr, arr + n);
    int left = 1, right = arr[n - 1] - arr[0];

    int ans = 0;
    while (left <= right) {
        int mid = (left + right) / 2;

        int cnt = c - 1, prev = arr[0];

        for (int i = 1; i < n; i++) {
            if (arr[i] - prev >= mid) {
                cnt--;
                prev = arr[i];
            }
        }
        if (cnt <= 0) {
            left = mid + 1;
            ans = max(ans, mid);
        } else {
            right = mid - 1;
        }
    }
    cout << ans << '\n';
    return 0;
}

File: BOJ/Binary_Search/BJ2110.js
const fs = require('fs');
const input = fs.readFileSync('/dev/stdin')
  .toString()
  .trim()
  .split('\n');

const [n, c] = input
  .shift()
  .split(' ')
  .map(Number);

const x = input
  .map(Number)
  .sort((a, b) => a - b);

let left = 0;
let right = 1_000_000_000;
let ans = 0;

while(left <= right) {
  const mid = Math.floor((left + right) / 2);

  let cnt = c - 1;
  let prev = x[0];
  for(let i = 1; i < n; i++) {
    if(x[i] - prev >= mid) {
      cnt--;
      prev = x[i];
    }
  }
  if(cnt > 0) {
    right = mid - 1;
  } else {
    left = mid + 1;
    ans = mid;
  }
}
console.log(ans);